chainstate/src/detail/mod.rs

```diff
@@ -408,56 +399,174 @@ impl<S: BlockchainStorage, V: TransactionVerificationStrategy> Chainstate<S, V>
+        let block_indices_to_invalidate = {
+            let chainstate_ref = self.make_db_tx_ro().map_err(BlockError::from).log_err()?;
+            assert!(!chainstate_ref
+                .is_block_in_main_chain(&(*block_id).into())
+                .map_err(BlockError::PropertyQueryError)
+                .log_err()?);
+
+            let block_index = get_existing_block_index(&chainstate_ref, block_id).log_err()?;
+            let next_block_height = get_next_block_height(block_index.block_height()).log_err()?;
+
+            let maybe_descendant_block_ids = chainstate_ref
+                .get_block_id_tree_top_as_list(next_block_height)
+                .map_err(BlockError::PropertyQueryError)?;
+
+            let mut block_indices_to_invalidate = Vec::new();
+            let mut seen_invalid_block_ids = BTreeSet::new();
+            block_indices_to_invalidate.push(block_index);
+            seen_invalid_block_ids.insert(*block_id);
+
+            for cur_block_id in maybe_descendant_block_ids {
+                let block_index =
+                    get_existing_block_index(&chainstate_ref, &cur_block_id).log_err()?;
+                let prev_block_id = block_index
+                    .prev_block_id()
+                    .classify(&self.chain_config)
+                    .chain_block_id()
+                    .expect("Genesis at non-zero height");
+
+                if seen_invalid_block_ids.contains(&prev_block_id) {
+                    block_indices_to_invalidate.push(block_index);
+                    seen_invalid_block_ids.insert(cur_block_id);
+                }
+            }
+
+            block_indices_to_invalidate
+        };
+
+        self.with_rw_tx(
+            |chainstate_ref| {
+                for (i, block_index) in block_indices_to_invalidate.iter().enumerate() {
+                    let mut status = block_index.status();
+                    if i == 0 {
+                        status.set_validation_failed()
+                    } else {
+                        status.set_has_invalid_parent();
+                    }
+
+                    let block_index = block_index.clone().with_status(status);
+                    chainstate_ref.set_block_index(&block_index)?;
+
+                    // FIXME: is it a good idea to remove blocks at this point though?
+                    // (note that because of this line some tests now run much longer than before).
+                    // Probably we should have a separate "garbage collecting" step that
+                    // would remove bad blocks (and probably outdated indices of bad blocks too).
+                    // In any case, block removal should be in a separate db transaction with
+                    // very few commit attempts (probably just 1) and delete_block's errors
+                    // should be ignored.
+                    chainstate_ref.delete_block(block_index.block_id())?;
```

Comment:
Invalidating blocks doesn't entail deleting them.

Ideally speaking, if you want to make it possible to delete blocks, the block index should be aware that blocks may not exist. Bitcoin Core has a status flag for that as well. A whole other story.

In my head, for simplicity, I was planning to keep the blocks and prune them later once we reach the reorg depth.

Reply:
> if you want to make it possible to delete blocks, the block index should be aware that blocks may not exist.

Just in case, we already can have this situation - when an invalid block gets added and is marked as invalid right away, we store a BlockIndex for it but don't store the block itself.

In general, should I turn this into a TODO and write an issue for now?